src/components/periodic-report/handlers/sync-periodic-report-to-project.handler.ts

```diff
@@ -0,0 +1,115 @@
+import { isNull, node, relation } from 'cypher-query-builder';
+import { DateTime } from 'luxon';
+import { Session } from '../../../common';
+import {
+  DatabaseService,
+  EventsHandler,
+  IEventHandler,
+  ILogger,
+  Logger,
+} from '../../../core';
+import { Project } from '../../project';
+import { ProjectUpdatedEvent } from '../../project/events';
+import { CreatePeriodicReport } from '../dto';
+import { ReportType } from '../dto/type.enum';
+import { PeriodicReportService } from '../periodic-report.service';
+
+type SubscribedEvent = ProjectUpdatedEvent;
+
+@EventsHandler(ProjectUpdatedEvent)
+export class SyncPeriodicReportToProject
+  implements IEventHandler<SubscribedEvent> {
+  constructor(
+    private readonly db: DatabaseService,
+    private readonly periodicReports: PeriodicReportService,
+    @Logger('periodicReport:sync-project') private readonly logger: ILogger
+  ) {}
+
+  async handle(event: SubscribedEvent) {
+    this.logger.debug('Project mutation, syncing periodic reports', {
+      ...event,
+      event: event.constructor.name,
+    });
+
+    const project = event.updated;
+    const previous = event.previous;
+
+    const isDateRangeChanged =
+      project.mouStart !== previous.mouStart ||
+      project.mouEnd !== previous.mouEnd;
+
+    if (isDateRangeChanged) {
+      await this.removeOldReports(project, event.session);
+      await this.syncRecords(project, event.session);
+    }
+  }
+
+  private async removeOldReports(project: Project, session: Session) {
+    const reports = await this.db
+      .query()
+      .match([
+        node('project', 'Project', { id: project.id }),
+        relation('out', '', 'report', { active: true }),
+        node('report', 'PeriodicReport'),
+      ])
+      .optionalMatch([
+        node('report'),
+        relation('out', 'rel', 'reportFile', { active: true }),
+        node('file', 'File'),
+      ])
+      .with('report, rel')
+      .where({ rel: isNull() })
+      .return('report.id as reportId')
+      .asResult<{ reportId: string }>()
+      .run();
+
+    await Promise.all(
+      reports.map((report) =>
+        this.periodicReports.delete(report.reportId, session)
+      )
+    );
+  }
+
+  private async syncRecords(project: Project, session: Session) {
+    await this.addReport(
+      project.id,
+      {
+        start: project.mouStart.value!,
+        end: project.mouEnd.value!,
+        type: ReportType.Financial,
+      },
+      session
+    );
```

Comment:
This will create a single report for the entire range of the project. We actually want multiple reports at a monthly/quarterly interval. Narrative & Progress reports can be hardcoded to quarterly, but financial needs to be looked up from the project property.

Reply:
@CarsonF This makes sense to me but if the project's date range is 2020-02-13 ~ 2020-06-29, how many progress/narrative reports should be created with what start/end date fields?

Reply:
I believe the date range should expand to the end of the quarter/month. If the project has been worked one week into the month, we still need the last report for that partial month.

Luxon has `endOf('month')`, etc. btw. I think quarter refers to our fiscal quarters though which are offset. I'll confirm with @sethmcknight.